File: 1335-minimum-difficulty-of-a-job-schedule/1335-minimum-difficulty-of-a-job-schedule.js
/**
 * @param {number[]} jobDifficulty
 * @param {number} d
 * @return {number}
 */
var minDifficulty = function(jobDifficulty, d) {
  const n = jobDifficulty.length

  if (d > n) {
    return -1
  }

  const dp = Array(n)
    .fill(null)
    .map(() => Array(d).fill(Infinity))

  dp[0][0] = jobDifficulty[0]

  for (let i = 1; i < n; i++) {
    dp[i][0] = Math.max(dp[i - 1][0], jobDifficulty[i])

    const jMax = Math.min(i + 1, d)
    for (let j = 1; j < jMax; j++) {
      let rangeMax = 0

      for (let k = i - 1; k >= 0; k--) {
        rangeMax = Math.max(jobDifficulty[k + 1], rangeMax)

        dp[i][j] = Math.min(dp[i][j], dp[k][j - 1] + rangeMax)
      }
    }
  }

  return dp[n - 1][d - 1]
};

File: 1335-minimum-difficulty-of-a-job-schedule/1335-minimum-difficulty-of-a-job-schedule.go
func minDifficulty(jobDifficulty []int, d int) int {
	n := len(jobDifficulty)

	if d > n {
		return -1
	}

	inf := 1000*300 + 1

	dp := make([][]int, n)
	for i := 0; i < n; i++ {
		dp[i] = make([]int, d)
		for j := 0; j < d; j++ {
			dp[i][j] = inf
		}
	}
	dp[0][0] = jobDifficulty[0]

	for i := 1; i < n; i++ {
		dp[i][0] = max(dp[i-1][0], jobDifficulty[i])

		for j := 1; j < min(i+1, d); j++ {
			rangeMax := 0

			for k := i - 1; k >= 0; k-- {
				rangeMax = max(rangeMax, jobDifficulty[k+1])

				dp[i][j] = min(dp[i][j], dp[k][j-1]+rangeMax)
			}
		}
	}

	return dp[n-1][d-1]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}